Homework4/DiagonalMatrix/MatrixTracker.cs

```diff
@@ -0,0 +1,32 @@
+﻿
+namespace DiagonalMatrix
+{
+    public class MatrixTracker<T>
+    {
+        int IndexI { get; set; }
+        int IndexJ { get; set; }
+        T OldValue { get; set; }
+        T NewValue { get; set; }
+        DiagonalMatrix<T> MatrixToTrack { get; }
+
+
+        public MatrixTracker(DiagonalMatrix<T> matrix)
+        {
+            matrix.ElementChanged += MatrixElementChanged;
+            MatrixToTrack = matrix; // To know which matrix needs Undo method.
+        }
+
+        private void MatrixElementChanged(int indexI, int indexJ, T oldValue, T newValue)
+        {
+            IndexI = indexI;
+            IndexJ = indexJ;
+            OldValue = oldValue;
+            NewValue = newValue;
+        }
+
+        public void Undo()
+        {
+            MatrixToTrack[IndexI, IndexJ] = OldValue;
```

Comment:
When no element have been changed, but Undo called, then [0,0] element is set to default(T).
Handle that case, restore element only when needed.